// rollup-plugin-postcss-modules/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = require("fs");
const path = __importStar(require("path"));
const camelcase_1 = __importDefault(require("camelcase"));
const rollup_plugin_postcss_1 = __importDefault(require("rollup-plugin-postcss"));
const reserved_words_1 = __importDefault(require("reserved-words"));
function fixname(name) {
    const ccName = camelcase_1.default(name);
    return reserved_words_1.default.check(ccName) ? `$${ccName}$` : ccName;
}
const formatCSSDefinition = (name, classNames) => `\
${classNames.filter((n) => !/-/.test(n)).map((t) => `export const ${t}: string`).join('\n')}
interface Namespace {
	${classNames.map((t) => `${JSON.stringify(t)}: string,`).join('\n\t')}
}
declare const ${name}: Namespace
export default ${name}`;
function writeCSSDefinition(cssPath, classNames) {
    return __awaiter(this, void 0, void 0, function* () {
        const name = fixname(path.basename(cssPath, '.css'));
        const definition = formatCSSDefinition(name, classNames);
        const dPath = `${cssPath}.d.ts`;
        yield fs_1.promises.writeFile(dPath, `${definition}\n`);
        return dPath;
    });
}
class CSSExports {
    constructor(writeDefinitions) {
        this.definitionCB = (dPath) => __awaiter(this, void 0, void 0, function* () {
            if (typeof this.writeDefinitions === 'function') {
                yield Promise.resolve(this.writeDefinitions(dPath));
            }
            else {
                console.log(`${dPath} written`);
            }
        });
        this.getJSON = (id, exportTokens) => __awaiter(this, void 0, void 0, function* () {
            const ccTokens = {};
            for (const className of Object.keys(exportTokens)) {
                ccTokens[fixname(className)] = exportTokens[className];
                ccTokens[className] = exportTokens[className];
            }
            if (this.writeDefinitions) {
                const dPath = yield writeCSSDefinition(id, Object.keys(ccTokens));
                yield this.definitionCB(dPath);
            }
        });
        this.writeDefinitions = writeDefinitions;
    }
}
function eslintPluginPostCSSModules(options = {}) {
    const { plugins = [], writeDefinitions = false, modules = {}, namedExports = fixname } = options, rest = __rest(options, ["plugins", "writeDefinitions", "modules", "namedExports"]);
    if ('getExport' in rest) {
        throw new Error("'getExport' is no longer supported.");
    }
    if (plugins.some((p) => p.postcssPlugin === 'postcss-modules')) {
        throw new Error("'rollup-plugin-postcss-modules' provides a 'postcss-modules' plugin, you cannot specify your own. Use the `modules` config key for configuration.");
    }
    const modulesOptions = modules === true ? {} : modules;
    if (modulesOptions === false || modulesOptions.getJSON) {
        throw new Error("'rollup-plugin-postcss-modules' provides a 'postcss-modules' plugin and its `getJSON()`. You cannot specify `modules.getJSON`");
    }
    const { getJSON } = new CSSExports(writeDefinitions);
    return rollup_plugin_postcss_1.default(Object.assign({ plugins: [...plugins], modules: Object.assign({ getJSON }, modulesOptions), autoModules: false, namedExports }, rest));
}
exports.default = eslintPluginPostCSSModules;
